Mathlib/Control/Random.lean

```diff
@@ -0,0 +1,145 @@
+import Mathlib.Init.Algebra.Order
+import Mathlib.Init.Data.Nat.Lemmas
+import Mathlib.Init.Data.Int.Order
+import Mathlib.Data.Fin.Basic
+import Mathlib.Data.Nat.Basic
+
+/-!
+# Rand Monad and Random Class
+This module provides tools for formulating computations guided by randomness and for
+defining objects that can be created randomly.
+## Main definitions
+  * `Rand` and `RandG` monad for computations guided by randomness;
+  * `Random` class for objects that can be generated randomly;
+    * `random` to generate one object;
+  * `BoundedRandom` class for objects that can be generated randomly inside a range;
+    * `randomR` to generate one object inside a range;
+  * `IO.runRand` to run a randomized computation inside the `IO` monad;
+## References
+  * Similar library in Haskell: https://hackage.haskell.org/package/MonadRandom
+-/
+
+/-- The RNGs that can be used to run a computation inside the `Rand`
+    Monad. Apart from generating random numbers they also have to be
+    capable of running inside IO like `StdGen`.
+-/
+inductive Rng where
+| stdGen : StdGen → Rng
+
+/-- A monad to generate random objects using the generic generator type `g` -/
+@[reducible]
+def RandG (g : Type) (α : Type u) := StateM (ULift g) α
+
+/-- A monad to generate random objects using the generator type `Rng` -/
+@[reducible]
+def Rand (α : Type u) := RandG Rng α
```

Comment:
```suggestion
abbrev Rand (α : Type u) := RandG Rng α
```

Reply:
I'm not sure I see the use case for the `Rng` type though.  You're always better of writing `RandG g` instead of `Rand` afaict.